src/mame/heathkit/tlb.cpp

```diff
@@ -0,0 +1,688 @@
+	{
+		uint8_t *decode = m_kbdrom->base();
+
+		m_transchar = decode[translate_mm5740_b(m_mm5740->b_r())];
+		m_strobe = true;
+		m_maincpu->set_input_line(INPUT_LINE_IRQ0, ASSERT_LINE);
+	}
+}
+
+MC6845_UPDATE_ROW(heath_tlb_device::crtc_update_row)
+{
+	if (!de)
+	{
+		return;
+	}
+
+	rgb_t const *const palette = m_palette->palette()->entry_list_raw();
+	uint32_t *p = &bitmap.pix(y);
+
+	for (uint16_t x = 0; x < x_count; x++)
+	{
+		uint8_t inv = (x == cursor_x) ? 0xff : 0;
+
+		uint8_t chr = m_p_videoram[(ma + x) & 0x7ff];
+
+		if (chr & 0x80)
+		{
+			inv ^= 0xff;
+			chr &= 0x7f;
+		}
+
+		// get pattern of pixels for that character scanline
+		uint8_t const gfx = m_p_chargen[(chr<<4) | ra] ^ inv;
+
+		// Display a scanline of a character (8 pixels)
+		*p++ = palette[BIT(gfx, 7)];
+		*p++ = palette[BIT(gfx, 6)];
+		*p++ = palette[BIT(gfx, 5)];
+		*p++ = palette[BIT(gfx, 4)];
+		*p++ = palette[BIT(gfx, 3)];
+		*p++ = palette[BIT(gfx, 2)];
+		*p++ = palette[BIT(gfx, 1)];
+		*p++ = palette[BIT(gfx, 0)];
+	}
+}
+
+
+// F4 Character Displayer
+static const gfx_layout h19_charlayout =
+{
+	8, 10,                  // 8 x 10 characters
+	128,                    // 128 characters
+	1,                      // 1 bits per pixel
+	{ 0 },                  // no bitplanes
+	// x offsets
+	{ 0, 1, 2, 3, 4, 5, 6, 7 },
+	// y offsets
+	{ 0*8, 1*8, 2*8, 3*8, 4*8, 5*8, 6*8, 7*8, 8*8, 9*8 },
+	8*16                    // every char takes 16 bytes
+};
+
+static GFXDECODE_START(gfx_h19)
+	GFXDECODE_ENTRY(":tlb:chargen", 0x0000, h19_charlayout, 0, 1)
+GFXDECODE_END
+
+
+/* Input ports */
+static INPUT_PORTS_START( tlb )
+
+	PORT_START("MODIFIERS")
+	// bit 0 connects to B8 of MM5740 - low if either shift key is
+	// bit 7 is low if a key is pressed
+	PORT_BIT(0x002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("CapsLock")   PORT_CODE(KEYCODE_CAPSLOCK) PORT_TOGGLE
+	PORT_BIT(0x004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Break")      PORT_CODE(KEYCODE_PAUSE)
+	PORT_BIT(0x008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("OffLine")    PORT_CODE(KEYCODE_F12)       PORT_TOGGLE
+	PORT_BIT(0x010, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("CTRL")       PORT_CODE(KEYCODE_LCONTROL)  PORT_CHAR(UCHAR_MAMEKEY(LCONTROL))
+	PORT_BIT(0x020, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("LeftShift")  PORT_CODE(KEYCODE_LSHIFT)    PORT_CHAR(UCHAR_SHIFT_1)
+	PORT_BIT(0x040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Repeat")     PORT_CODE(KEYCODE_LALT)
+	PORT_BIT(0x100, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("RightShift") PORT_CODE(KEYCODE_RSHIFT)
+	PORT_BIT(0x200, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Reset")      PORT_CODE(KEYCODE_F10)
+
+	PORT_START("X1")
+	PORT_BIT(0x001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("/")          PORT_CODE(KEYCODE_SLASH)      PORT_CHAR('/') PORT_CHAR('?')
+	PORT_BIT(0x002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Unused")
+	PORT_BIT(0x004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Unused")
+	PORT_BIT(0x008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Unused")
```

Comment:
The unused positions in the matrix should really be `IPT_UNUSED` – they shouldn’t show up to the user.  I know they’re useful during development, but you shouldn’t leave them in.